substrate-archive/src/wasm_tracing.rs

```diff
@@ -0,0 +1,332 @@
+	fn to_string(&self) -> String {
+		match self {
+			DataType::Bool(b) => b.to_string(),
+			DataType::I64(i) => i.to_string(),
+			DataType::U64(u) => u.to_string(),
+			DataType::String(s) => s.to_string(),
+		}
+	}
+}
+
+impl From<DataType> for String {
+	fn from(data: DataType) -> String {
+		data.to_string()
+	}
+}
+
+#[derive(Default, Debug, Clone, Serialize, Deserialize)]
+pub struct TraceData(HashMap<String, DataType>);
+
+impl Visit for TraceData {
+	fn record_debug(&mut self, field: &Field, value: &dyn std::fmt::Debug) {
+		self.0.insert(field.name().to_string(), DataType::String(format!("{:?}", value)));
+	}
+
+	fn record_i64(&mut self, field: &Field, value: i64) {
+		self.0.insert(field.name().to_string(), DataType::I64(value));
+	}
+
+	fn record_u64(&mut self, field: &Field, value: u64) {
+		self.0.insert(field.name().to_string(), DataType::U64(value));
+	}
+
+	fn record_str(&mut self, field: &Field, value: &str) {
+		self.0.insert(field.name().to_string(), DataType::String(value.to_string()));
+	}
+
+	fn record_bool(&mut self, field: &Field, value: bool) {
+		self.0.insert(field.name().to_string(), DataType::Bool(value));
+	}
+}
+
+impl Subscriber for TraceHandler {
+	fn enabled(&self, metadata: &Metadata<'_>) -> bool {
+		self.targets.iter().any(|(t, _l)| metadata.target().starts_with(t.as_str()))
+	}
+
+	fn new_span(&self, attrs: &Attributes<'_>) -> Id {
+		let meta = attrs.metadata();
+		let mut values = TraceData::default();
+		attrs.record(&mut values);
+
+		let id = Id::from_u64(self.counter.fetch_add(1, Ordering::Relaxed));
+		let span_message = SpanMessage {
+			id: id.clone(),
+			parent_id: attrs.parent().cloned().or_else(|| self.current_span.id()),
+			name: meta.name().to_string(),
+			target: meta.target().to_string(),
+			level: *meta.level(),
+			start_time: Utc::now(),
+			overall_time: chrono::Duration::zero(),
+			file: None,
+			line: None,
+			values,
+		};
+
+		if self.is_enabled(&span_message) {
+			self.gather_span(span_message).unwrap_or_else(|e| log::error!("{}", e.to_string()));
+		}
+
+		id
+	}
+
+	fn record(&self, id: &Id, values: &Record<'_>) {
+		if let Some(span) = self.span_events.lock().spans.iter_mut().find(|span| &span.id == id) {
+			values.record(&mut span.values);
+		}
+	}
+
+	fn record_follows_from(&self, _: &Id, _: &Id) {
+		log::warn!("Followed span relationship not recorded");
+	}
+
+	fn event(&self, event: &Event<'_>) {
+		let time = Utc::now();
+		if let Err(e) = self.gather_event(event, time) {
+			log::error!("{}", e.to_string());
+		}
+	}
+
+	fn enter(&self, id: &Id) {
+		self.current_span.enter(id.clone());
+	}
+
+	fn exit(&self, _id: &Id) {
+		self.current_span.exit();
+	}
+
+	fn try_close(&self, id: Id) -> bool {
+		let end_time = Utc::now();
+		if let Some(span) = self.span_events.lock().spans.iter_mut().find(|span| &span.id == &id) {
+			span.overall_time = end_time - span.start_time;
+		}
+		// try_close returns false by default -- we want to keep this behavior.
```

Comment:
Not actually sure about this -- we don't track span references by id or anything, so I figured that doing what the default impl does is the most sensible thing here. It works on my end for the DB. But not sure if returning `false` unilaterally here means that we can re-enter the span again sort of like in `exit`. @mattrutherford ?

by using this does this mean we have to also track clone_span to make sure all spans are dropped?